File: julia/src/interpolation/newton_i.jl
module NewtonI

export newton_i

"""
        newton_i(x_nodes, y_nodes, x)

Newton's divided difference interpolation.
Returns the interpolated value P(x).
"""

function newton_i(x_nodes::Vector{<:Real}, y_nodes::Vector{<:Real}, x::Real)
    n = length(x_nodes)
    n == length(y_nodes) || error("x_nodes and y_nodes must have the same length.")

    coef = copy(y_nodes)

    for j in 1:n-1
        for i in n:-1:j+1
            coef[i] = (coef[i] - coef[i-1]) / (x_nodes[i] - xnodes[i-j])
        end
    end

    result = coef[n]
    for k in (n-1):-1:1
        result = result * (x - x_nodes[k]) + coef[k]
    end

    return result
end

if abspath(PROGRAM_FILE) == @__FILE__
    xs = [0.0, 1.0, 2.0]
    ys = [1.0, 3.0, 2.0]
    println(newton_i(xs, ys, 1.5))
end

end

File: fortran/src/interpolation/newton_i.f90
module newton_mod
  implicit none
contains

  function newton_i(x_nodes, y_nodes, x, n) result(p)
    implicit none
    integer, intent(in) :: n
    real(8), intent(in) :: x_nodes(n), y_nodes(n), x
    real(8) :: p
    real(8) :: coef(n)
    integer :: i, j

    coef = y_nodes

    do j = 2, n
       do i = n, j, -1
          coef(i) = (coef(i) - coef(i-1)) / (x_nodes(i) - x_nodes(i-j+1))
       end do
    end do

    p = coef(n)
    do i = n-1, 1, -1
       p = p * (x - x_nodes(i)) + coef(i)
    end do

  end function newton_i

end module newton_mod
